src/backend/cuda/kernel/random_engine.hpp

```diff
@@ -0,0 +1,655 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+#include <dispatch.hpp>
+#include <err_cuda.hpp>
+#include <debug_cuda.hpp>
+#include <af/defines.h>
+#include <kernel/random_engine_philox.hpp>
+#include <kernel/random_engine_threefry.hpp>
+#include <kernel/random_engine_mersenne.hpp>
+#include <random_engine.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+    //Utils
+
+    static const int THREADS = 256;
+    #define UINTMAXFLOAT 4294967296.0f
+    #define UINTLMAXDOUBLE (4294967296.0*4294967296.0)
+    #define PI_VAL 3.1415926535897932384626433832795028841971693993751058209749445923078164
+
+    __device__ static float getFloat(const uint &num)
+    {
+        return float(num)/UINTMAXFLOAT;
+    }
+
+    __device__ static double getDouble(const uint &num1, const uint &num2)
+    {
+        uintl num = (((uintl)num1)<<32) | ((uintl)num2);
+        return double(num)/UINTLMAXDOUBLE;
+    }
+
+    template <typename T>
+    __device__ static void boxMullerTransform(T * const out1, T * const out2, const T &r1, const T &r2)
+    {
+        T r = sqrt((T)(-2.0) * log(r1));
```

Comment:
Also, I'm sure it's in the algorithm, but still the sqrt(**-2.0**) looks really weird.